File: gateway/app/core/cache.py
import hashlib
import logging
from functools import lru_cache

from redis import Redis
from redis.exceptions import RedisError

from app.core.config import settings

logger = logging.getLogger(__name__)


def _blacklist_key(token: str) -> str:
    return f"user-service:blacklist:{hashlib.sha256(token.encode('utf-8')).hexdigest()}"


@lru_cache(maxsize=1)
def get_redis_client() -> Redis | None:
    if not settings.redis_enabled:
        return None
    return Redis(
        host=settings.redis_host,
        port=settings.redis_port,
        db=settings.redis_db,
        password=settings.redis_password or None,
        decode_responses=True,
        socket_connect_timeout=1,
        socket_timeout=1,
    )


def _safe_execute(operation: str, func):
    client = get_redis_client()
    if client is None:
        return None
    try:
        return func(client)
    except RedisError as exc:
        logger.warning("Redis %s failed: %s", operation, exc)
        return None


def is_token_blacklisted(token: str) -> bool:
    value = _safe_execute("check blacklist", lambda client: client.get(_blacklist_key(token)))
    return value is not None


def increment_rate_limit(key: str, window_seconds: int) -> int | None:
    def _increment(client: Redis):
        pipe = client.pipeline()
        pipe.incr(key)
        pipe.ttl(key)
        current, ttl = pipe.execute()
        if ttl == -1:
            client.expire(key, window_seconds)
        elif ttl == -2:
            client.expire(key, window_seconds)
        return int(current)

    return _safe_execute("increment rate limit", _increment)

File: gateway-go/internal/cache/cache.go
package cache

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"log"
	"time"

	"github.com/redis/go-redis/v9"

	"modern-user-product-order-system/gateway-go/internal/config"
)

type RedisClient interface {
	Get(ctx context.Context, key string) *redis.StringCmd
	Incr(ctx context.Context, key string) *redis.IntCmd
	TTL(ctx context.Context, key string) *redis.DurationCmd
	Expire(ctx context.Context, key string, expiration time.Duration) *redis.BoolCmd
}

type Cache struct {
	client RedisClient
}

func New(cfg config.Config) *Cache {
	if !cfg.RedisEnabled {
		return &Cache{}
	}
	client := redis.NewClient(&redis.Options{
		Addr:         fmt.Sprintf("%s:%d", cfg.RedisHost, cfg.RedisPort),
		Password:     cfg.RedisPassword,
		DB:           cfg.RedisDB,
		DialTimeout:  time.Second,
		ReadTimeout:  time.Second,
		WriteTimeout: time.Second,
	})
	return &Cache{client: client}
}

func NewWithClient(client RedisClient) *Cache {
	return &Cache{client: client}
}

func BlacklistKey(token string) string {
	sum := sha256.Sum256([]byte(token))
	return "user-service:blacklist:" + hex.EncodeToString(sum[:])
}

func (c *Cache) IsTokenBlacklisted(ctx context.Context, token string) bool {
	if c.client == nil {
		return false
	}
	value, err := c.client.Get(ctx, BlacklistKey(token)).Result()
	if err == redis.Nil {
		return false
	}
	if err != nil {
		log.Printf("Redis check blacklist failed: %v", err)
		return false
	}
	return value != ""
}

func (c *Cache) IncrementRateLimit(ctx context.Context, key string, windowSeconds int) *int {
	if c.client == nil {
		return nil
	}
	current, err := c.client.Incr(ctx, key).Result()
	if err != nil {
		log.Printf("Redis increment rate limit failed: %v", err)
		return nil
	}
	ttl, err := c.client.TTL(ctx, key).Result()
	if err != nil {
		log.Printf("Redis rate limit ttl failed: %v", err)
		return nil
	}
	if ttl < 0 {
		if err := c.client.Expire(ctx, key, time.Duration(windowSeconds)*time.Second).Err(); err != nil {
			log.Printf("Redis rate limit expire failed: %v", err)
			return nil
		}
	}
	value := int(current)
	return &value
}
